### PR TITLE
New Observation service

### PR DESCRIPTION
**Pull request by BarisYazici**
- Observation service has its own function
- It returns q, qdot and feasibility attribute.

### CODE REVIEW

File: src/robots/msj_platform.cpp [hunk 1]
```diff
@@ -128,7 +129,7 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
         motor_command.publish(msg);
     };
 
-    int pnpoly(vector<double> limits_x, vector<double> limits_y, double testx, double testy){
+    int isFeasible(vector<double> limits_x, vector<double> limits_y, double testx, double testy){
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp]: great!

Comment (tomasruizt) [inline src/robots/msj_platform.cpp:132]: great!

File: src/robots/msj_platform.cpp [hunk 2]
```diff
@@ -203,33 +203,41 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
         }
     }
 
+    ///Return only qdot, q and feasibility to gym environment
+    bool GymObservationService(roboy_simulation_msgs::GymStep::Request &req,
+                        roboy_simulation_msgs::GymStep::Response &res){
+
+        setResponse(q,qd,res);
+        if(isFeasible(limits[0],limits[1],q[0],q[1])){
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp]: lets simplify with:
`res.feasible = isFeasible(...);`

File: src/robots/msj_platform.cpp [hunk 3]
```diff
@@ -203,33 +203,41 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
...
+            //Set the commanded tendon velocity from RL agent to simulation
+            Ld[0][i] = req.set_points[i];
+        }
+        if(!external_robot_state)
+            forwardKinematics(req.step_size);
+
+        ROS_INFO_STREAM_THROTTLE(5, "Ld = " << Ld[0].format(fmt));
+        write();
+        if(isFeasible(limits[0],limits[1],q[0],q[1])){
             res.feasible = true;
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp]: lets simplify here as well `res.feasible = isFeasble(...);`

File: src/robots/msj_platform.cpp [hunk 4]
```diff
@@ -203,33 +203,41 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
...
-	            forwardKinematics(req.step_size);
-
-	        ROS_INFO_STREAM_THROTTLE(5, "Ld = " << Ld[0].format(fmt));
-
-	        write();
-	    }
-        if(pnpoly(limits[0],limits[1],q[0],q[1])){
-            //task space is feasible
+
+        update();
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp:217]: Have you tried setting the tendons before calling update()? Makes more sense here

File: src/robots/msj_platform.cpp [hunk 5]
```diff
@@ -203,33 +203,41 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
...
+        for(int i=0; i< number_of_cables; i++){
+            //Set the commanded tendon velocity from RL agent to simulation
+            Ld[0][i] = req.set_points[i];
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp:220]: Shall we rename `set_points` with a more meaningful name? If you agree we can create an Issue for the next time we make changes on the messages.

File: src/robots/msj_platform.cpp [hunk 6]
```diff
@@ -203,33 +203,41 @@ class MsjPlatform: public cardsflow::kindyn::Robot{
...
+        }
+        if(!external_robot_state)
+            forwardKinematics(req.step_size);
+
+        ROS_INFO_STREAM_THROTTLE(5, "Ld = " << Ld[0].format(fmt));
+        write();
+        if(isFeasible(limits[0],limits[1],q[0],q[1])){
```
Comment (tomasruizt) [inline src/robots/msj_platform.cpp]: lets simplify here as well `res.feasible = isFeasble(...);`

**Review by tomasruizt — CHANGES_REQUESTED** [review]

File: src/robots/msj_platform.cpp (hunk 4 above)
Comment (BarisYazici) [reply src/robots/msj_platform.cpp:217]: I tried. It doesn't work with that configuration

File: src/robots/msj_platform.cpp (hunk 5 above)
Comment (BarisYazici) [reply src/robots/msj_platform.cpp:220]: We can put this into our service message type PR

**Pull request merged**